### PR TITLE
SNOW-799408: Close temp file handle before removing temp cache file

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-799408

temp file is already an open file handle, instead of opening it again, we directly write the bytes into the file handle and then close the fd

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: test/extras/run.py [hunk 1]
```diff
@@ -23,3 +25,17 @@
             ]
         )
         sub_process.check_returncode()
+        ocsp_cache_dir_path = pathlib.Path(
+            snowflake.connector.ocsp_snowflake.OCSP_RESPONSE_VALIDATION_CACHE.file_path
+        ).parent
+        cache_files = set(os.listdir(ocsp_cache_dir_path))
+        # This is to test SNOW-79940, making sure tmp files are removed
+        # Windows does not have ocsp_response_validation_cache.lock
```
Comment (sfc-gh-stan) [inline test/extras/run.py:33]: Let's explicitly detect the platform and assert separately for windows in an if case.

File: src/snowflake/connector/cache.py [hunk 2]
```diff
@@ -524,7 +524,16 @@ def _save(self, load_first: bool = True) -> bool:
...
-                    with open(tmp_file, "wb") as w_file:
+                    # tmp_file is an opened OS level handle, which means we need to close it manually.
+                    # https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp
+                    # ideally we shall just use the tmp_file fd to write,
+                    # however, using os.write(tmp_file, bytes) causes seg fault during garbage collection when exiting
+                    # python program.
+                    # thus we fall back to the approach using the normal open() method to open a file and write.
+                    os.close(tmp_file)
+                    with open(tmp_file_path, "wb") as w_file:
+                        # note: during garbage collection when exiting python program, open will raise not defined
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/cache.py:535]: I'm a bit confused here. What do you mean by "we depend on this behavior to exit the program gracefully"?

File: test/extras/run.py (hunk 1 above)
Comment (sfc-gh-aling) [reply test/extras/run.py:33]: smart, I will make the change

File: src/snowflake/connector/cache.py [hunk 3]
```diff
@@ -524,7 +524,16 @@ def _save(self, load_first: bool = True) -> bool:
...
                     )
-                    with open(tmp_file, "wb") as w_file:
+                    # tmp_file is an opened OS level handle, which means we need to close it manually.
+                    # https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp
+                    # ideally we shall just use the tmp_file fd to write,
+                    # however, using os.write(tmp_file, bytes) causes seg fault during garbage collection when exiting
+                    # python program.
+                    # thus we fall back to the approach using the normal open() method to open a file and write.
+                    os.close(tmp_file)
+                    with open(tmp_file_path, "wb") as w_file:
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:534]: I fail to see how this is different from the previous code. The docs [here](https://docs.python.org/3/library/functions.html#open) say:

> (If a file descriptor is given, it is closed when the returned I/O object is closed unless closefd is set to False.)

So to me it sounds like file descriptor, or path. The I/O object returned by `open` should behave identically.

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/cache.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:534]: you are right about the usage of open.

the problem is when exiting the program, during garbage collection, the `open` method is garbage collected, so calling `with open` would raise error, so it didn't get the chance the close the underlying fd.

without `os.close`, the status of a fd during garbage collection is:
opened by mktmpfile -> error raised when calling `open` -> left opened

with `os.close`, the status:
opened -> closed by os.close ->   error raised when calling `open` -> left closed

Comment (sfc-gh-mkeller) [reply src/snowflake/connector/cache.py:534]: If we want to protect against this specific scenario, what about doing:
```
try:
	with open(tmp_file, "wb") as w_file:
		...
except AppropiateException:
	# If file descriptor is still open close it
	try:
		os.fstat(tmp_file)
		os.close(tmp_file)
	except OSError:
		pass
```

**Pull request merged**